support torch script call over rpc

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30063 support torch script call over rpc**

Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

zhaojuanmao committed:
create a torch script rpc API



Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao commented:
Not for review yet

zhaojuanmao committed:
Update on "create a torch script rpc API"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/script_call.cpp:
```diff
@@ -13,10 +13,35 @@ ScriptCall::ScriptCall(
+    : qualifiedName_(qualifiedName), stack_(args) {}
+
+bool ScriptCall::hasOp() const {
+  if (op_) {
```
nit: just do `return op_`? Same for `hasQualifiedName` below.

zhaojuanmao replied:
will clean it up

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -284,6 +286,46 @@ If the future completes with an error, an exception is thrown.
+    c10::intrusive_ptr<c10::ivalue::Future> fut;
+  };
+
+  py::class_<PythonFutureWrapper>(module, "PythonFutureWrapper")
```
this should be bind to "Future" instead of "PythonFutureWrapper" right?

zhaojuanmao replied:
right, good catch

wanchaol commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -284,6 +286,46 @@ If the future completes with an error, an exception is thrown.
         return pyRpcPythonUdf(agent, dst, pickledPythonUDF, tensors);
       });
 
+  struct PythonFutureWrapper {
```
can you add a TODO here saying this should be properly merged with the jit PythonFutureWrapper as one of the task when merging these two future implementations?

zhaojuanmao replied:
will do

pritamdamania87 replied:
Would be nice to create a gh issue for this follow up too.

kostmo commented:
## CircleCI build failures summary
As of commit 685e282:
* **2/2** broken upstream at merge base 409151e (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=409151e1bb90c66546901d372f59930f72242ee9&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=685e2827caf4a7d2232fbff67b238c52184e221c).

### 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75317283)
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75320204)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

wanchaol commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -49,7 +49,15 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
       auto& stack = scriptCall.stackRef();
-      scriptCall.op()->getOperation()(stack);
+      at::IValue res;
+      if (scriptCall.hasOp()) {
```
can you provide the ScriptCall class a callable overload operator, and wrap these logics in there? This might clean the logic and apis (i.e. `hasOp` and `hasQualifiedName` is confusing, we could just run the call directly).

zhaojuanmao replied:
will do

zhaojuanmao replied:
tried to move the logic to script call class, then the script call class will have python dependency, this results in that all the files depending on script call class have to move from C++ torch lib to python torch libs. I feel it is a little overkilled. 

So still keeping the logic here

wanchaol commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -0,0 +1,53 @@
+      *agent, agent->getWorkerInfo(dst), std::move(*scriptCall).toMessage());
+  // Get function return type to construct c10::ivalue::Future.
+  // Script call only allows single IValue returned.
+  auto returns = PythonRpcHandler::getInstance()
```
`ScriptCall` should just record the return type that we get from the compilation unit.

wanchaol commented on torch/csrc/distributed/rpc/script_functions.h:
```diff
@@ -0,0 +1,25 @@
+//                  "moduleName::fnName", e.g, "dist_autograd_test::my_py_add"
+//   stack: a bags of IValue args passed to fnName
+// It returns IValue that is c10::intrusive_ptr<ivalue::Future>
+c10::IValue rpcTorchscriptCall(
```
nit:TorchScriptCall

zhaojuanmao replied:
want to differentiate from remoteTorchScriptCall in the future

wanchaol commented on torch/csrc/distributed/rpc/script_remote_call.cpp:
```diff
@@ -47,9 +47,9 @@ std::unique_ptr<ScriptRemoteCall> ScriptRemoteCall::fromMessage(
+  auto scriptCallPtr = ScriptCall::fromIValues(values);
   return c10::guts::make_unique<ScriptRemoteCall>(
-      op, std::move(values), std::move(retRRefId), std::move(retForkId));
+      scriptCallPtr->op(), std::move(values), retRRefId, retForkId);
```
GlobalUniqueIds should use move semantics to avoid extra copying? why we remove here?

zhaojuanmao replied:
ah, possibly an accidental change, will change it back

zhaojuanmao commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -0,0 +1,53 @@
+      "Return value of a annotated torchScript function should be a single "
+      "IValue, got a vector of size ",
+      returns.size());
+  auto returnType = returns.at(0).type();
```
return type is here

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -1228,6 +1255,40 @@ def test_backward_simple_python_udf(self):
+                local_grads = ret if ret else local_grads
+
+        # Right now rpc torchscript call does not accept annotated torchscript
+        # class name or script module class name or their class method names
```
Can we create gh issues to track adding support for these later?

pritamdamania87 commented on torch/csrc/distributed/rpc/script_call.cpp:
```diff
@@ -58,9 +90,11 @@ std::shared_ptr<Operator> ScriptCall::fromIValues(
+    return std::make_unique<ScriptCall>(op, std::move(ivalues));
   } else {
-    TORCH_CHECK(false, "Unrecognized qualified name ", qualifiedName);
+    ivalues.pop_back();
```
nit: ivalues.pop_back() can be outside the if-else block and be shared by both codepaths?

pritamdamania87 commented on torch/csrc/distributed/rpc/script_functions.h:
```diff
@@ -0,0 +1,25 @@
+//                  "moduleName::fnName", e.g, "dist_autograd_test::my_py_add"
+//   stack: a bags of IValue args passed to fnName
+// It returns IValue that is c10::intrusive_ptr<ivalue::Future>
+c10::IValue rpcTorchscriptCall(
```
Shouldn't the return type for this be ivalue::Future for clarity?

pritamdamania87 commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -0,0 +1,53 @@
+                              const rpc::Message& message,
+                              const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      throw std::runtime_error((*futErr).what());
```
Don't think we should throw an exception in the callback. Shouldn't we call `markCompleted(FutureError&&)` here instead?

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -78,6 +79,23 @@ def my_rref_add(rref_t1, t2):
+@torch.jit.script
+def my_script_add(t1, t2):
+    return torch.add(t1, t2)
+
```
Is it possible to write a torchscript function that throws an exception and validate that the client receives the appropriate exception? For example the body of the function could call torch operators with invalid args.

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -87,7 +89,7 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
     }
     case MessageType::BACKWARD_AUTOGRAD_RESP: {
-      return autograd::RpcWithAutograd::fromMessage(response);
+      return autograd::PropagateGradientsResp::fromMessage(response);
```
Do you know why none of our unit tests caught this issue before? Do we have a unit test covering this now?

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -99,6 +101,39 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
+      // Handle the original RPC.
+      auto wrappedMessageType = rpcWithAutograd.wrappedMessageType();
+      return deserializeResptoIValueInternal(
+          rpcWithAutograd.wrappedRpc(), wrappedMessageType);
```
Can we move the autograd logic to `deserializeResponse`? That way python_functions.cpp and this method reuse the same logic instead of repeating it in both places?

Looks like `deserializeResponse` doesn't unwrap `FORWARD_AUTOGRAD_RESP`, we can change it to unwrap it and call `deserializeResponse` recursively. That way we share this logic in one place and simplifies our logic here to just checking whether the resp type is SCRIPT_RET.

zhaojuanmao replied:
good idea!

zhaojuanmao replied:
@pritamdamania87 I has a few tries, and can not get deserializeResponse() refactored cleanly if moving FORWARD_AUTOGRAD_RESP unwrap logic into deserializeResponse(), the reasons are following: 
1) FORWARD_AUTOGRAD_RESP message can only return RpcCommandBase& for wrappedRpc
2) deserializeResponse() needs to return wrappedMessageType, otherwise no way to know type of wrappedRpc

So we need interface like this: 

RpcCommandBase& deserializeResponse(const message& response, unique_ptr<RpcCommandBase>& rpc, MessageType& msgType)  // rpc unique_ptr is used to keep returned RpcCommandBase& alive

The users have to write codes like this:
unique_ptr<RpcCommandBase> rpc;
MessageType msgType = message.type();
auto& unwrappedRpc = deserializeResponse(message, rpc, msgType)
deserializeIValue(unwrappedRpc, msgType)

The above change does not sound clean to me. I prefer to keep deserializeResponse() simple as it is right now, how do you think?

pritamdamania87 replied:
As discussed offline, lets move the `wrappedRpc_` out of RpcWithAutograd and add a messageType field to RpcCommandBase.

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.cpp:
```diff
@@ -99,6 +101,39 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
+          rpcWithAutograd.wrappedRpc(), wrappedMessageType);
+    }
+    default: {
+      return IValue();
```
shouldn't we throw an exception here since this is an invalid state?

pritamdamania87 commented on test/dist_autograd_test.py:
```diff
@@ -1228,6 +1255,40 @@ def test_backward_simple_python_udf(self):
+                if ExecMode.LOCAL == exec_mode:
+                    ret = self._exec_func(exec_mode, my_script_add, t1, t2)
+                else:
+                    ret = self._exec_func(exec_mode, _qualified_name(my_script_add), t1, t2)
```
Looks like the user needs to call `_qualified_name(func)` right now? Can we make it such that you can just pass `func` into the API and we call `_qualified_name(func)` internally?

pritamdamania87 commented on torch/distributed/rpc/api.py:
```diff
@@ -382,3 +383,98 @@ def rpc_async(to, func, args=None, kwargs=None):
+    """
+    args = args if args else ()
+    kwargs = kwargs if kwargs else {}
+    fut = _invoke_rpc_script(to, qualified_name, *args, **kwargs)
```
I was wondering if we can avoid having a separate API for torchscript call. Is it possible to do this in our current RPC API?

```
try:
   qname = torch.jit._qualified_name(func)
   return _invoke_rpc_script(...)
except:
   return _invoke_rpc(...)
```

pritamdamania87 reviewed (commented):
Awesome work!

mrshenli commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -263,7 +271,7 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
     }
   }
-}
+} // namespace rpc
```
isn't this closing bracket for `processRpc` instead of a namespace?

mrshenli commented on torch/csrc/distributed/rpc/script_functions.h:
```diff
@@ -0,0 +1,25 @@
+namespace distributed {
+namespace rpc {
+
+// This function sends a rpc call to run torchscript function, currently the
```
a rpc -> an rpc

mrshenli commented on torch/csrc/distributed/rpc/script_functions.h:
```diff
@@ -0,0 +1,25 @@
+//   dst: destination worker name
+//   qualifiedName: torchscript function qualified name string like
+//                  "moduleName::fnName", e.g, "dist_autograd_test::my_py_add"
+//   stack: a bags of IValue args passed to fnName
```
a bags -> a bag
and what is fnName here? Do you mean TorchScript function?

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -382,3 +383,98 @@ def rpc_async(to, func, args=None, kwargs=None):
+        >>> import torch.distributed.rpc as rpc
+        >>> dist.init_process_group(backend='gloo', rank=0, world_size=2)
+        >>> rpc.init_model_parallel("worker0")
+        >>> ret = rpc.rpc_sync("worker1", "myModule::myTorchScriptFunc", args=(torch.ones(2), 3))
```
Let's also include an example of `myModule::myTorchScriptFunc` here.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -284,6 +286,54 @@ If the future completes with an error, an exception is thrown.
+
+  module.def(
+      "_invoke_rpc_script",
+      [](const std::string& dst,
```
Any reason for supporting string-only dst instead of a WorkerInfo?

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -382,3 +383,98 @@ def rpc_async(to, func, args=None, kwargs=None):
+@_require_initialized
+def _rpc_sync(to, qualified_name, args=None, kwargs=None):
+    r"""
+    Make a blocking RPC call to run function ``func`` on worker ``to``. RPC
```
Let's say it explicitly that this only works for TorchScript functions.

mrshenli reviewed (commented):
This is great!! Thanks for adding this!

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao commented:
@pritamdamania87 and @mrshenli I addressed your comments except two unresolved ones, please see inline comments. Would you please taking another look at them? thanks for reviewing!

pritamdamania87 reviewed (changes requested):
Requesting changes for the deserializeResponse refactor and torchscript support in the public API as we discussed offline.

zhaojuanmao commented on test/dist_autograd_test.py:
```diff
@@ -1195,6 +1222,40 @@ def test_backward_simple_python_udf(self):
+        # Right now rpc torchscript call does not accept annotated torchscript
+        # class name or script module class name or their class method names
+        with self.assertRaisesRegex(RuntimeError, "attempted to get undefined function"):
+            ret = rpc._rpc_sync('worker{}'.format(self._next_rank()),
```
yes

zhaojuanmao commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -41,11 +41,13 @@ c10::IValue rpcTorchscriptCall(
     if (futErr) {
-      throw std::runtime_error((*futErr).what());
+      c10::ivalue::Future::FutureError jitFutErr(std::string((*futErr).what()));
+      futPtr->markCompleted(std::move(jitFutErr));
```
futPtr is ivalue::Future here, it only has markCompleted(error) right now

mrshenli commented:
@zhaojuanmao  Two minor comments:

1. As this is going to enhance the scope of the public API, let's update the docs too with an example explaining how to use it.
2. Could you please also update the PR/diff description with some more details, as it will show up in git log. Thanks!

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao commented:
addressed comments. for deserializeResponse() refactoring, I did not add messageType to RpcCommandBase, as it does not work. we need to know messageType before casting RpcCommandBase to derived Rpc, but  before casting, we can not know messageType from base class RpcCommandBase. To  make it work, I still let deserializeResponse() return a wrappedMessageType by reference. 

would you please take another look at it? thanks

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/autograd/rpc_messages/rpc_with_autograd.h:
```diff
@@ -42,6 +42,8 @@ class TORCH_API RpcWithAutograd final : public rpc::RpcCommandBase {
 
   RpcCommandBase& wrappedRpc();
 
+  std::unique_ptr<RpcCommandBase>&& moveWrappedRpc() &&;
```
isn't `std::unique_ptr<RpcCommadBase>` sufficient? Why do we a rvalue reference to the unique pointer itself?

zhaojuanmao replied:
right, let me change that

zhaojuanmao commented:
@pytorchbot retest this please

pritamdamania87 commented on torch/csrc/distributed/rpc/utils.h:
```diff
@@ -14,7 +14,8 @@ TORCH_API std::unique_ptr<RpcCommandBase> deserializeRequest(
 TORCH_API std::unique_ptr<RpcCommandBase> deserializeResponse(
-    const Message& response);
+    const Message& response,
+    MessageType& wrappedMsgType);
```
nit: Can we enhance the docs to mention what `wrappedMsgType` refers to? Also, can we make `wrappedMsgType` be an optional argument and default to `UNKNOWN`? Then in the `FORWARD_AUTOGRAD_RESP` case we assert that it is not unknown? (since its used only there).

zhaojuanmao replied:
will add doc. and let me try to change wrappedMsgType default to UNKNOWN

zhaojuanmao replied:
hmm, wrappedMsgType is non const reference, it can not be bind to temp default value, so I'm keeping this as it is right now

pritamdamania87 reviewed (approved):
Looks great! Thanks a lot for the nice refactor in deserializeResponse! Although, before landing make sure you update the PR summary as @mrshenli mentioned. Also, I have a few comments inline, please look through those as well.

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "support torch script call over rpc"


Differential Revision: [D18482934](https://our.internmc.facebook.com/intern/diff/D18482934/)

[ghstack-poisoned]

suo commented:
I think this broke master :( Reverting

zhaojuanmao commented:
Wait, let me check the log for py35, and the previous commit before mine has not finish master test on py35 yet?

From: Michael Suo <notifications@github.com>
Reply-To: pytorch/pytorch <reply@reply.github.com>
Date: Tuesday, January 14, 2020 at 1:26 PM
To: pytorch/pytorch <pytorch@noreply.github.com>
Cc: Yanli Zhao <yanlizhao@fb.com>, Mention <mention@noreply.github.com>
Subject: Re: [pytorch/pytorch] support torch script call over rpc (#30063)


I think this broke master :( Reverting

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/30063?email_source=notifications&email_token=ALTZIOXZKGF2UAWEYECIQHDQ5YUWTA5CNFSM4JO4WXR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI6F2UI#issuecomment-574381393>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALTZIOTIPKM2ZIN6NOUVFGDQ5YUWTANCNFSM4JO4WXRQ>.

suo commented:
The test that is failing was introduced by this PR, so I felt confident enough to revert it :( Sorry for the churn if I was wrong :/

zhaojuanmao commented:
Np, I just checked the log, yes it is related. Thanks!

From: Michael Suo <notifications@github.com>
Reply-To: pytorch/pytorch <reply@reply.github.com>
Date: Tuesday, January 14, 2020 at 1:36 PM
To: pytorch/pytorch <pytorch@noreply.github.com>
Cc: Yanli Zhao <yanlizhao@fb.com>, Mention <mention@noreply.github.com>
Subject: Re: [pytorch/pytorch] support torch script call over rpc (#30063)


The test that is failing was introduced by this PR, so I felt confident enough to revert it :( Sorry for the churn if I was wrong :/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/30063?email_source=notifications&email_token=ALTZIOUWVTBEM34B4GBGSLLQ5YVYXA5CNFSM4JO4WXR2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI6GU5Y#issuecomment-574384759>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALTZIOQJFE7IBYM2PJU26FDQ5YVYXANCNFSM4JO4WXRQ>.

zhaojuanmao commented:
The issue is: Py35 throws error message like "Can't pickle ..", and Py36 throws error message like "can't pickle....", and one unit test expects error message to match "can't pickle....", so the unit test failed in Py35 tests. 

will add [test all] tag before re-landing

mrshenli commented:
@xush6528 just discovered that, if something crashed and prevents us from running `_cleanup_python_rpc_handler()` it will crash and only crash in PY35. It might have sth to do with the destruction order problem we encounter before. Hope this might be helpful for debugging.

zhaojuanmao commented:
@mrshenli py35 failure here is not related to python_rpc_handler, it is because my newly added test tries to greps and match error message, and py35 and py36 error messages have different styles.
